File: MangaViewer/app/src/main/java/com/king/mangaviewer/model/MangaChapterItem.java
package com.king.mangaviewer.model;

import android.text.TextUtils;

import com.king.mangaviewer.util.StringUtils;

public class MangaChapterItem extends BaseItem {
    MangaMenuItem menu = null;

    /**
     * @param id
     * @param title
     * @param description
     * @param imagePath
     * @param menu
     */

    public MangaChapterItem(String id, String title, String description,
                            String imagePath, String url, MangaMenuItem menu) {
        super(id, title, description, imagePath, url, menu.getMangaWebSource());
        this.menu = menu;
    }


    public MangaMenuItem getMenu() {
        return menu;
    }


    public void setMenu(MangaMenuItem menu) {
        this.menu = menu;
    }

    public String getHash() {
        StringBuilder sb = new StringBuilder();
        sb.append(menu.getHash() + "|" + this.getUrl());
        return StringUtils.getHash(sb.toString());
    }

    @Override
    public String getTitle() {
        String title = super.getTitle();
        if (!TextUtils.isEmpty(title)) {
            return title.replace(getMenu().getTitle(), "");
        }
        return super.getTitle();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof MangaChapterItem) {
            return this.getHash().equals(((MangaChapterItem) obj).getHash());
        }
        return super.equals(obj);
    }
}

File: MangaViewer/app/src/main/java/com/king/mangaviewer/model/MangaChapterItem.kt
package com.king.mangaviewer.model

import android.text.TextUtils

import com.king.mangaviewer.util.StringUtils

open class MangaChapterItem(
    var id: String,
    var title: String,
    var description: String,
    var imagePath: String,
    var url: String,
    var menu: MangaMenuItem) {

    var hash: String = StringUtils.getHash(url.trim())
    val mangaWebSource: MangaWebSource
        get() = menu.mangaWebSource

    init {
        title = if (!TextUtils.isEmpty(title)) {
            title.replace(menu.title + "", "")
        } else title
    }


    override fun equals(other: Any?): Boolean {
        return if (other is MangaChapterItem) {
            this.hash == other.hash
        } else super.equals(other)
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + title.hashCode()
        result = 31 * result + description.hashCode()
        result = 31 * result + imagePath.hashCode()
        result = 31 * result + url.hashCode()
        result = 31 * result + menu.hashCode()
        result = 31 * result + hash.hashCode()
        return result
    }
}
